src/shims/env.rs

```diff
@@ -68,20 +69,20 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
         name_op: OpTy<'tcx, Tag>,
         value_op: OpTy<'tcx, Tag>,
     ) -> InterpResult<'tcx, i32> {
-        let this = self.eval_context_mut();
+        let mut this = self.eval_context_mut();
 
         let name_ptr = this.read_scalar(name_op)?.not_undef()?;
         let value_ptr = this.read_scalar(value_op)?.not_undef()?;
-        let value = this.memory.read_c_str(value_ptr)?;
+        let value = this.read_os_str_from_c_str(value_ptr)?;
         let mut new = None;
         if !this.is_null(name_ptr)? {
-            let name = this.memory.read_c_str(name_ptr)?;
-            if !name.is_empty() && !name.contains(&b'=') {
+            let name = this.read_os_str_from_c_str(name_ptr)?;
+            if !name.is_empty() && !os_str_to_bytes(&name).unwrap().contains(&b'=') {
                 new = Some((name.to_owned(), value.to_owned()));
             }
         }
         if let Some((name, value)) = new {
-            let var_ptr = alloc_env_var(&name, &value, &mut this.memory);
+            let var_ptr = alloc_env_var_as_c_str(name.as_os_str(), value.as_os_str(), &mut this);
```

Comment:
```suggestion
            let var_ptr = alloc_env_var_as_c_str(&name, &value, &mut this);
```
I think this should also work.

Reply:
Thank you! Confirmed working in my local build 👍